### PR TITLE
Array comparison and sourceVals.

### PR DESCRIPTION
**Pull request by sbarzowski**
This change introduces the ability to compare
arrays lexicographically using the customary operator (<, >, <=, >=).

This change needs to be followed up by appropriate changes in the docs
and go-jsonnet implementation.

To make this change, I introduced "sourceVals" – the values from std
which can be used relatively easily in the interpreter. In particular
the operator< is implemented as jsonnet function __array_less which
is called by the interpreter.

This is much less efficient than a builtin implementation of course,
but implmentation of builtins here tends to be really painful and
this allows us to quickly implement features which would otherwise
be very tricky. Also if performance is needed, go-jsonent should be used
anyway (where builtin implemetation requires easily 10x less effort).

### CODE REVIEW

File: core/vm.cpp [hunk 1]
```diff
@@ -881,8 +921,18 @@ class Interpreter {
         builtins["parseJson"] = &Interpreter::builtinParseJson;
         builtins["encodeUTF8"] = &Interpreter::builtinEncodeUTF8;
         builtins["decodeUTF8"] = &Interpreter::builtinDecodeUTF8;
+
+        DesugaredObject *stdlib = makeStdlibAST(alloc, "__internal__");
+        jsonnet_static_analysis(stdlib);
+        stdlibAST = stdlib; // stdlibAST is const, so we need to do analysis before this assignment
+        stdThunk = makeHeap<HeapThunk>(nullptr, nullptr, 0, static_cast<const AST*>(stdlibAST));
```
Comment (sparkprime) [inline core/vm.cpp:928]: Can stdThunk be a local var instead of a field?  And is it possible to do it without the thunk at all?

Comment (sparkprime) [reply core/vm.cpp:928]: I think if you let the context be null in the newCall, the worst that will happen is a slightly worse stacktrace if there was a problem evaluating the stdlib (which is a simple file containing only an object literal anyway).

**Review by sparkprime — APPROVED** [review]
Is sourceVals actually simpler than desugaring the < to a call like with % and std.format?

File: test_suite/array_comparison2.jsonnet [hunk 2]
```diff
@@ -0,0 +1,18 @@
+local arrs = [
+  [[]],
+  [[1, 2]],
+  [[1], [2]],
+  [[1, 2], []],
+];
+[
+  {
+    'x < y': x < y,
```
[the remark this answers is not in the document]
Comment (sbarzowski) [reply test_suite/array_comparison2.jsonnet]: Done

File: stdlib/std.jsonnet [hunk 3]
```diff
@@ -1379,4 +1379,38 @@ limitations under the License.
...
+  __compare_array(arr1, arr2)::
+    local len1 = std.length(arr1), len2 = std.length(arr2);
+    local minLen = std.min(len1, len2);
+    local aux(i) =
+      if i < minLen then
+        local cmpRes = std.__compare(arr1[i], arr2[i]);
+        if cmpRes != 0 then
+          cmpRes
+        else
+          aux(i + 1)
```
[the remark this answers is not in the document]
Comment (sbarzowski) [reply stdlib/std.jsonnet]: Good idea, done.

**Comment by sbarzowski** [comment]
> Is sourceVals actually simpler than desugaring the < to a call like with % and std.format?

No it's not. Maybe it would be if we had all the setup for it ready.

I considered desugaring, but I was afraid it would butcher performance. With sourceVals we can pay for it only when we need it. Also, going forward, it might be a useful crutch for keeping semantic parity between implementations.

**Pull request merged**